Add custom filter

charleskawczynski opened the pull request:
### Description

The EDMF team has demonstrated significantly more stable simulations using custom filters. cc: @yairchn @ilopezgp 

The current implementation of the custom filter is a Q&D hack that takes many passes over the entire MPIStateArray:
```julia
cbtmarfilter = GenericCallbacks.EveryXSimulationSteps(1) do
    bl = solver_config.dg.balance_law
    st = vars_state(bl, Prognostic(), eltype(Q))
    aux = vars_state(bl, Auxiliary(), eltype(Q))
    a_min = bl.turbconv.subdomains.a_min
    a_max = bl.turbconv.subdomains.a_max
    i_ρ = varsindex(st, :ρ)
    i_ρq_tot_gm = varsindex(st, :moisture, :ρq_tot)
    ρ_gm = solver_config.Q[:,i_ρ,:]
    i_ρaθ_liq_en = varsindex(aux, :turbconv, :environment,:ρaθ_liq)
    ρaθ_liq_en  = solver_config.dg.state_auxiliary[:,i_ρaθ_liq_en,:]
    ρaθ_liq_ups = FT(0)
    ρaq_tot_ups = FT(0)
    ρa_ups      = FT(0)
    ρaw_ups     = FT(0)
    for i in 1:n_updrafts(bl.turbconv)
        i_ρaθ_liq_up = varsindex(st, :turbconv, :updraft, Val(i), :ρaθ_liq)
        i_ρaq_tot_up = varsindex(st, :turbconv, :updraft, Val(i), :ρaq_tot)
        i_ρa_up      = varsindex(st, :turbconv, :updraft, Val(i), :ρa)
        i_ρaw_up     = varsindex(st, :turbconv, :updraft, Val(i), :ρaw)
        ρaθ_liq_ups  = ρaθ_liq_ups .+ solver_config.Q[:,i_ρaθ_liq_up,:]
        ρaq_tot_ups  = ρaq_tot_ups .+ solver_config.Q[:,i_ρaq_tot_up,:]
        ρa_ups       = ρa_ups .+ solver_config.Q[:,i_ρa_up,:]
        ρaw_ups      = ρaw_ups .+ solver_config.Q[:,i_ρaw_up,:]
    end
    ρq_tot_gm   = solver_config.Q[:,i_ρq_tot_gm,:]
    ρaw_en      = - ρaw_ups
    ρaq_tot_en  = (ρq_tot_gm - ρaq_tot_ups) ./ (ρ_gm .- ρa_ups)
    θ_liq_en    = ρaθ_liq_en ./ (ρ_gm .- ρa_ups)
    q_tot_en    = ρaq_tot_en ./ (ρ_gm .- ρa_ups)
    w_en        = ρaw_en./ (ρ_gm .- ρa_ups)
    for i in 1:n_updrafts(bl.turbconv)
        i_ρa_up      = varsindex(st, :turbconv, :updraft, Val(i), :ρa)
        i_ρaθ_liq_up = varsindex(st, :turbconv, :updraft, Val(i), :ρaθ_liq)
        i_ρaq_tot_up = varsindex(st, :turbconv, :updraft, Val(i), :ρaq_tot)
        i_ρaw_up     = varsindex(st, :turbconv, :updraft, Val(i), :ρaw)
        a_up_mask   = solver_config.Q[:,i_ρa_up,:] .< ρ_gm*a_min
        ρ_area_change = max.(a_up_mask.*(ρ_gm*a_min .- solver_config.Q[:,i_ρa_up,:]),FT(0))
        solver_config.Q[:,i_ρa_up,:]      .+= a_up_mask .*ρ_area_change
        solver_config.Q[:,i_ρaθ_liq_up,:] .+= a_up_mask .*θ_liq_en.*ρ_area_change
        solver_config.Q[:,i_ρaq_tot_up,:] .+= a_up_mask .*q_tot_en.*ρ_area_change
        solver_config.Q[:,i_ρaw_up,:]     .+= a_up_mask .*w_en.*ρ_area_change
    end
    nothing
end
```

With this PR, we can simplify this to a single pass:

```julia
import ClimateMachine.DGMethods: custom_filter!
struct EDMFFilter <: AbstractCustomFilter end
function custom_filter!(::EDMFFilter, atmos, state, aux)
    FT = eltype(state)
    bl = atmos
    a_min = bl.turbconv.subdomains.a_min
    a_max = bl.turbconv.subdomains.a_max
    en = state.turbconv.environment
    up = state.turbconv.updraft
    N_up = n_updrafts(bl.turbconv)
    ρ_gm = state.ρ
    ρaθ_liq_ups = sum(ntuple(up[i].ρaθ_liq, N_up))
    ρaq_tot_ups = sum(ntuple(up[i].ρaq_tot, N_up))
    ρa_ups      = sum(ntuple(up[i].ρa, N_up))
    ρaw_ups     = sum(ntuple(up[i].ρaw, N_up))
    ρq_tot_gm   = state.ρq_tot
    ρaw_en      = - ρaw_ups
    ρaq_tot_en  = (ρq_tot_gm - ρaq_tot_ups) / (ρ_gm - ρa_ups)
    θ_liq_en    = en.ρaθ_liq / (ρ_gm - ρa_ups)
    q_tot_en    = ρaq_tot_en / (ρ_gm - ρa_ups)
    w_en        = ρaw_en / (ρ_gm - ρa_ups)
    @unroll_map(N_up) do i
        a_up_mask   = up[i].ρa < ρ_gm*a_min
        ρ_area_change = max(a_up_mask * (ρ_gm * a_min - up[i].ρa), FT(0))
        up[i].ρa      += a_up_mask * ρ_area_change
        up[i].ρaθ_liq += a_up_mask * θ_liq_en * ρ_area_change
        up[i].ρaq_tot += a_up_mask * q_tot_en * ρ_area_change
        up[i].ρaw     += a_up_mask * w_en * ρ_area_change
    end
end

cbtmarfilter = GenericCallbacks.EveryXSimulationSteps(1) do
    apply!(
        EDMFFilter(),
        solver_config.dg.grid,
        solver_config.dg.balance_law,
        solver_config.Q,
        solver_config.dg.state_auxiliary,
    )
    nothing
end
```

I'm aware that this filter will not preserve mass, however, I think that either 1) we may be able to orthogonally improve that or 2) it may not matter based on the assumptions made in the EDMF model.

<!-- Check all the boxes below before taking the PR out of draft -->

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

jkozdon commented on src/Numerics/DGMethods/custom_filter.jl:
```diff
@@ -0,0 +1,93 @@
+    elems,
+    activedofs,
+) where {dim, N}
+    FT = eltype(state_prognostic)
+    num_state_prognostic = number_states(bl, Prognostic())
+    num_state_auxiliary = number_states(bl, Auxiliary())
+    vs_p = Vars{vars_state(bl, Prognostic(), FT)}
+    vs_a = Vars{vars_state(bl, Auxiliary(), FT)}
+    Nq = N + 1
+    Nqk = dim == 2 ? 1 : Nq
+    Np = Nq * Nq * Nqk
+
+    local_state_prognostic = MArray{Tuple{num_state_prognostic}, FT}(undef)
+    local_state_auxiliary = MArray{Tuple{num_state_auxiliary}, FT}(undef)
```
```suggestion
begin @uniform
    FT = eltype(state_prognostic)
    num_state_prognostic = number_states(bl, Prognostic())
    num_state_auxiliary = number_states(bl, Auxiliary())
    vs_p = Vars{vars_state(bl, Prognostic(), FT)}
    vs_a = Vars{vars_state(bl, Auxiliary(), FT)}
    Nq = N + 1
    Nqk = dim == 2 ? 1 : Nq
    Np = Nq * Nq * Nqk

    local_state_prognostic = MArray{Tuple{num_state_prognostic}, FT}(undef)
    local_state_auxiliary = MArray{Tuple{num_state_auxiliary}, FT}(undef)
end
```

jkozdon commented on test/Numerics/DGMethods/runtests.jl:
```diff
@@ -9,4 +9,5 @@ include(joinpath("..", "..", "testhelpers.jl"))
     runmpi(joinpath(@__DIR__, "integral_test.jl"))
     runmpi(joinpath(@__DIR__, "integral_test_sphere.jl"))
     runmpi(joinpath(@__DIR__, "vars_test.jl"))
+    runmpi(joinpath(@__DIR__, "custom_filter.jl"))
```
Test on GPU too?

jkozdon commented on test/Numerics/DGMethods/custom_filter.jl:
```diff
@@ -0,0 +1,81 @@
+    aux::Vars,
+    localgeo,
+)
+    state.q = 0
```
```suggestion
    state.q = hypot(localgeo.coord)
```

jkozdon commented on test/Numerics/DGMethods/custom_filter.jl:
```diff
@@ -0,0 +1,81 @@
+            state::Vars,
+            aux::Vars,
+        )
+            state.q = 1
```
```suggestion
            state.q = state.q^2
```

jkozdon commented on test/Numerics/DGMethods/custom_filter.jl:
```diff
@@ -0,0 +1,81 @@
+                state_gradient_flux = nothing,
+            )
+
+            Q = ClimateMachine.DGMethods.init_ode_state(dg)
```
```suggestion
            Q = ClimateMachine.DGMethods.init_ode_state(dg)
            data = Array(Q.realdata)
```

jkozdon commented on test/Numerics/DGMethods/custom_filter.jl:
```diff
@@ -0,0 +1,81 @@
+                CustomFilter(),
+            )
+
+            @test all(Q.realdata .≈ 1)
```
```suggestion
            @test all(Array(Q.realdata) .== data.^2)
```

jkozdon reviewed (commented):
Suggested a slightly better test.

I think this needs to be tested on the GPU (not sure it would work right now).

charleskawczynski commented:
> Suggested a slightly better test.... I think this needs to be tested on the GPU (not sure it would work right now).

I added your suggestions and added a GPU test to buildkite.

charleskawczynski commented:
bors try

charleskawczynski commented:
bors try

jkozdon commented on src/Numerics/DGMethods/custom_filter.jl:
```diff
@@ -0,0 +1,117 @@
+"""
+    AbstractCustomFilter <: AbstractFilter
+
+Apply a custom filter, `f!`, to the prognostic state.
```
```suggestion
Dispatch type for a pointwise custom filter; see [`custom_filter!`](@ref)
```

jkozdon reviewed (approved):
Looks good

charleskawczynski commented:
bors r+

charleskawczynski commented:
bors r+
